### PR TITLE
etd robots more decoupled from fedora 3

### PR DESCRIPTION
**Pull request by ndushay**
## Why was this change made?

Moves us a little further towards decoupling ETD robots from Fedora 3.   From https://github.com/sul-dlss/common-accessioning/issues/582#issuecomment-589858935: 

>completely decoupling the etd robots will not be possible until we modify hydra_etd to use the deposit endpoint. The etd robots depend on some non-standard datastreams (properties for one) that don't map to cocina models."

I'm not sure if this was part of (or all of) #582 or #586 - @jcoyne?

NOTE:  I am not confident the test code did a great job of ensuring that my changes are correct so careful review of my changes is welcome.  Or merge it and let it break.  whatever.

Also note:  I'm more inclined to do follow on PRs if this is incomplete in some way.

## Was the usage documentation (e.g. README, DevOpsDocs, wiki, queue specific README) updated?

n/a

### CODE REVIEW

File: .rubocop_todo.yml [hunk 1]
```diff
@@ -32,7 +32,7 @@ Lint/UnderscorePrefixedVariableName:
 
 # Offense count: 22
 Metrics/AbcSize:
-  Max: 98
+  Max: 100
```
Comment (jcoyne) [inline .rubocop_todo.yml]: 😱

File: lib/robots/dor_repo/etd_submit/other_metadata.rb [hunk 2]
```diff
@@ -11,51 +11,51 @@ def initialize(opts = {})
...
-          attrs = { mimeType: 'application/xml', dsLabel: label, content: metadata }
-          datastream = ActiveFedora::Datastream.new(etd.inner_object, ds_name, attrs)
-          datastream.controlGroup = 'M'
-          datastream.save
+        def create_legacy_metadata(druid, new_metadata)
+          object_client = Dor::Services::Client.object(druid)
+
+          # yes, legacy_update will create the metadata
+          object_client.metadata.legacy_update(
+            content: {
```
Comment (jcoyne) [inline lib/robots/dor_repo/etd_submit/other_metadata.rb]: This is the name of the datastream. In this case the content metadata. This should vary per datastream.

Comment (ndushay) [reply lib/robots/dor_repo/etd_submit/other_metadata.rb]: d'oh!

File: .rubocop_todo.yml (hunk 1 above)
Comment (ndushay) [reply .rubocop_todo.yml]: yeah, i know.  :-P

**Comment by ndushay** [comment]
thanks for catching the error!  I have fixed it and added a test.

File: lib/robots/dor_repo/etd_submit/check_marc.rb [hunk 3]
```diff
@@ -16,7 +16,8 @@ def initialize(opts = {})
 
         #========= This method check symphony and looks for MARC record based on druid ==========#
         def perform(druid)
-          etd = Etd.find(druid)
+          object_client = Dor::Services::Client.object(druid)
+          etd = object_client.find
```
Comment (jcoyne) [inline lib/robots/dor_repo/etd_submit/check_marc.rb]: It looks to me like this value gets passed to `Etd::IdentityMetadataGenerator.generate`, which requires an ActiveFedora Etd object.

File: lib/robots/dor_repo/etd_submit/other_metadata.rb [hunk 4]
```diff
@@ -11,51 +11,55 @@ def initialize(opts = {})
           super('etdSubmitWF', 'other-metadata', opts)
         end
 
-        # create metadata datastreams for the work item
+        # create metadata for the work item
         def perform(druid)
-          etd = Etd.find(druid)
+          object_client = Dor::Services::Client.object(druid)
+          etd = object_client.find
```
Comment (jcoyne) [inline lib/robots/dor_repo/etd_submit/other_metadata.rb]: It looks to me like this value gets passed to the *MetadataGenerators, which depend on an ActiveFedora Etd.

File: lib/robots/dor_repo/etd_submit/other_metadata.rb [hunk 5]
```diff
@@ -11,51 +11,55 @@ def initialize(opts = {})
...
-          attrs = { mimeType: 'application/xml', dsLabel: label, content: metadata }
-          datastream = ActiveFedora::Datastream.new(etd.inner_object, ds_name, attrs)
-          datastream.controlGroup = 'M'
-          datastream.save
+        def create_legacy_metadata(druid, md_sym, new_metadata)
+          object_client = Dor::Services::Client.object(druid)
+
+          # legacy_update will create the metadata
+          object_client.metadata.legacy_update(
+            md_sym => {
```
Comment (jcoyne) [inline lib/robots/dor_repo/etd_submit/other_metadata.rb]: This looks good. one optimization here is that we can generate all of the different content and pass them in a single `legacy_update` request.  e.g.:

```ruby
object_client.metadata.legacy_update(
  content: {
    updated: Time.now,
    content: new_metadata
  },
  identity: {
    ...
  },
  version: {
    ...
  },
  rights: {
    ...
  })

See http://sul-dlss.github.io/dor-services-app/#operation/metadata#update_legacy_metadata
```

File: spec/robots/etd_submit/check_marc_spec.rb [hunk 6]
```diff
@@ -46,6 +45,7 @@
       end
 
       before do
+        allow(object_client).to receive(:find).and_return(object)
```
Comment (jcoyne) [inline spec/robots/etd_submit/check_marc_spec.rb]: Dor::Services::Client::Object.find doesn't return a ActiveFedora Etd.

File: lib/robots/dor_repo/etd_submit/submit_marc.rb [hunk 7]
```diff
@@ -67,12 +67,13 @@ def output_file(druid, record)
         end
 
         def generate_marc(druid)
-          etd = Etd.find(druid)
+          object_client = Dor::Services::Client.object(druid)
+          etd = object_client.find
```
Comment (jcoyne) [inline lib/robots/dor_repo/etd_submit/submit_marc.rb]: These two calls are not equivalent, they return different classes of object.

File: spec/robots/etd_submit/submit_marc_spec.rb [hunk 8]
```diff
@@ -57,20 +57,20 @@ def cleanup
...
-      it "creates marc and saves it to a file in the day's working directory" do
-        skip
-        # marc_file = File.join(ROBOT_ROOT, 'tmp', Time.now.strftime('%Y%m%d'), "#{@mock_workitem.druid.gsub(':','_')}.marc")
-        #         FileUtils.rm(marc_file) if File.exists(marc_file)
-        #
-        #         @robot.process_item(@mock_workitem)
-        #         File.should be_file(marc_file)
-      end
-    end
+    # describe '#process_items' do
```
Comment (jcoyne) [inline spec/robots/etd_submit/submit_marc_spec.rb]: Just remove this?

File: lib/robots/dor_repo/etd_submit/check_marc.rb (hunk 3 above)
Comment (ndushay) [reply lib/robots/dor_repo/etd_submit/check_marc.rb]: so this should be left as `Etd.find()`?  or switch to `Dor::Etd.find` or `Dor::Item.find` or whatever?

File: lib/robots/dor_repo/etd_submit/other_metadata.rb (hunk 4 above)
Comment (ndushay) [reply lib/robots/dor_repo/etd_submit/other_metadata.rb]: same question as above.

File: lib/robots/dor_repo/etd_submit/other_metadata.rb (hunk 5 above)
Comment (ndushay) [reply lib/robots/dor_repo/etd_submit/other_metadata.rb]: will update

File: spec/robots/etd_submit/submit_marc_spec.rb (hunk 8 above)
Comment (ndushay) [reply spec/robots/etd_submit/submit_marc_spec.rb]: sure.

File: .rubocop_todo.yml (hunk 1 above)
Comment (ndushay) [reply .rubocop_todo.yml]: back to 98!!!  I didn't do it!!!!

File: .rubocop_todo.yml [hunk 9]
```diff
@@ -148,7 +149,7 @@ RSpec/ScatteredLet:
   Exclude:
     - 'spec/robots/assembly/jp2_create_spec.rb'
 
-# Offense count: 6
+# Offense count: 3
```
Comment (ndushay) [inline .rubocop_todo.yml]: look!  I made it better!  🤣

**Pull request merged**